### PR TITLE
[0.21] Speedy trial activation for Taproot

### PR DESCRIPTION
**Pull request by achow101**
Backport of #21377 and #21686

### CODE REVIEW

File: src/test/versionbits_tests.cpp [hunk 1]
```diff
@@ -226,6 +267,13 @@ BOOST_AUTO_TEST_CASE(versionbits_test)
         // Make sure that no deployment tries to set an invalid bit.
         BOOST_CHECK_EQUAL(bitmask & ~(uint32_t)VERSIONBITS_TOP_MASK, bitmask);
 
+        // Check min_activation_height is on a retarget boundary
+        BOOST_CHECK_EQUAL(mainnetParams.vDeployments[i].min_activation_height % mainnetParams.nMinerConfirmationWindow, 0);
+        // Check min_activation_height is 0 for ALWAYS_ACTIVE and never active deployments
+        if (mainnetParams.vDeployments[i].nStartTime == Consensus::BIP9Deployment::ALWAYS_ACTIVE || mainnetParams.vDeployments[i].nStartTime == Consensus::BIP9Deployment::NEVER_ACTIVE) {
+            BOOST_CHECK_EQUAL(mainnetParams.vDeployments[i].min_activation_height, 0);
```
Comment (MarcoFalke) [inline src/test/versionbits_tests.cpp:274]:
```suggestion
            BOOST_CHECK_EQUAL(mainnetParams.vDeployments[i].min_activation_height, 0U);
```

Looks like you are hitting some stupid compiler error

Comment (achow101) [reply src/test/versionbits_tests.cpp]: Fixed.

**Comment by Sjors** [comment]
tACK 724e4948004c8127f093db90f9d78e6ed9e46541

I verified as follows:

```
git checkout origin/0.21
git checkout -b speedy-trial-backport
git cherry-pick 63879f0a4760c0c0f784029849cb5d21ee088abb^..ffe33dfbd4c3b11e3475b022b6c1dd077613de79
git cherry-pick f979b3237f1cfc28f9c4ccb07beab558d5357a55
git diff 724e4948004c8127f093db90f9d78e6ed9e46541
src/test/test_bitcoin --run_test=versionbits_tests
````

Git diff only shows a difference in the trailing `U` for this test: `BOOST_CHECK_EQUAL(mainnetParams.vDeployments[i].min_activation_height, 0U);`

I had manually resolve a conflict for dd07e6da48040dc7eae46bc7941db48d98a669fd.

I also managed to spin up the node and see the softfork as "defined" in `getblockchaininfo`.

**Comment by achow101** [comment]
I don't know why CI is having a hard time building this. I am able to do a gitian build of this branch, so it should be working.

File: src/test/versionbits_tests.cpp [hunk 2]
```diff
@@ -226,6 +267,13 @@ BOOST_AUTO_TEST_CASE(versionbits_test)
         // Make sure that no deployment tries to set an invalid bit.
         BOOST_CHECK_EQUAL(bitmask & ~(uint32_t)VERSIONBITS_TOP_MASK, bitmask);
 
+        // Check min_activation_height is on a retarget boundary
+        BOOST_CHECK_EQUAL(mainnetParams.vDeployments[i].min_activation_height % mainnetParams.nMinerConfirmationWindow, 0);
```
Comment (fanquake) [inline src/test/versionbits_tests.cpp:271]:
```suggestion
        BOOST_CHECK_EQUAL(mainnetParams.vDeployments[i].min_activation_height % mainnetParams.nMinerConfirmationWindow, 0U);
```

File: src/test/versionbits_tests.cpp [hunk 3]
```diff
@@ -226,6 +267,13 @@ BOOST_AUTO_TEST_CASE(versionbits_test)
...
+        // Check min_activation_height is 0 for ALWAYS_ACTIVE and never active deployments
+        if (mainnetParams.vDeployments[i].nStartTime == Consensus::BIP9Deployment::ALWAYS_ACTIVE || mainnetParams.vDeployments[i].nStartTime == Consensus::BIP9Deployment::NEVER_ACTIVE) {
+            BOOST_CHECK_EQUAL(mainnetParams.vDeployments[i].min_activation_height, 0U);
```
Comment (fanquake) [inline src/test/versionbits_tests.cpp:274]:
```suggestion
            BOOST_CHECK_EQUAL(mainnetParams.vDeployments[i].min_activation_height, 0);
```

File: src/test/versionbits_tests.cpp [hunk 4]
```diff
@@ -242,81 +290,118 @@ BOOST_AUTO_TEST_CASE(versionbits_test)
...
+    if (nStartTime == Consensus::BIP9Deployment::NEVER_ACTIVE) return;
 
-    assert(nStartTime < nTimeout);
+    BOOST_REQUIRE(nStartTime < nTimeout);
+    BOOST_REQUIRE(nStartTime >= 0);
+    BOOST_REQUIRE(nTimeout <= std::numeric_limits<uint32_t>::max() || nTimeout == Consensus::BIP9Deployment::NO_TIMEOUT);
+    BOOST_REQUIRE(0 <= bit && bit < 32);
+    BOOST_REQUIRE(((1 << bit) & VERSIONBITS_TOP_MASK) == 0);
+    BOOST_REQUIRE(min_activation_height >= 0);
+    BOOST_REQUIRE_EQUAL(min_activation_height % params.nMinerConfirmationWindow, 0);
```
Comment (fanquake) [inline src/test/versionbits_tests.cpp:317]:
```suggestion
    BOOST_REQUIRE_EQUAL(min_activation_height % params.nMinerConfirmationWindow, 0U);
```

**Review by fanquake — COMMENTED** [review]
These 3 changes will fix the failing CIs.

File: src/test/versionbits_tests.cpp (hunk 2 above)
Comment (achow101) [reply src/test/versionbits_tests.cpp:271]: Done

File: src/test/versionbits_tests.cpp (hunk 3 above)
Comment (achow101) [reply src/test/versionbits_tests.cpp:274]: Done

File: src/test/versionbits_tests.cpp (hunk 4 above)
Comment (achow101) [reply src/test/versionbits_tests.cpp:317]: Done

**Comment by instagibbs** [comment]
cherry-pick ACK https://github.com/bitcoin/bitcoin/pull/21701/commits/cbd64c3a28a7466f421477daadc6e6e6b69b898a

**Comment by jnewbery** [comment]
ACK cbd64c3a28a7466f421477daadc6e6e6b69b898a

Repeated the rebase/cherry-pick myself and the only differences were the unsigned/signed fixups here: https://github.com/bitcoin/bitcoin/pull/21701#pullrequestreview-637244766.

**Comment by Sjors** [comment]
tACK cbd64c3

**Comment by MarcoFalke** [comment]
cherry-pick-only ACK cbd64c3a28a7466f421477daadc6e6e6b69b898a 🌾

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

cherry-pick-only ACK cbd64c3a28a7466f421477daadc6e6e6b69b898a 🌾
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiVXAwAm+m81VxCGAQs/vB2ZvE43rHYlW8QBpQbRB4ckkjnmXxSoNJipajP02Je
3CwjtSrxS9A0Dam8u4D3BHHdYyyT/UJF02JJMNEzTlO9b2OW5v0SrGZqGGIxSluC
oltxBDg5ry3tHSs4Zkm2o71je9L00ZgOMQD4Yan5yOCVmCVWfIaBFR3oXhL/pdhr
0UrlkPWPuF3h2Lt7dqEqhd3/rZzV7dtMK0tBqX7B7IE8sX/g5qvEIAAnstn6jJeN
rYbTwEkVMKDstIvonGa1EcMAr4TMay2V1h+q53JNvAmqkIj+8nZvH/5hbpnVNuTg
CivIvZKffc3C+4cigKsiJSHcT9vAeK6K/C5MW5G/9uio1kqdZq7McA3my7AmlDN5
e3jtTBDIUN/BMhOiizS1iYsqwHZjF4qPdfySxgDPw4Z49toee2ziN0BMv9NTrKc6
WI5hPKFG38l8O4YHO6tyCRJfVJuoP2ByhGiU7/pTJO0NvKnRh+Wma3kRa61sTbrl
cfriU22f
=gYxK
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `1d1c8419f375b140d9bb77c1f31aa2bb76a54e095fe73d7ba926604fb50abc88  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401081d1c8419f375b140d9bb77c1f31aa2bb76a54e095fe73d7ba926604fb50abc88f0108bcec35bb42f033500defb6bd9d561e108fff0107ac2f0a2d8b1769a208a2c365253b78608f104607954e8f00826bda2fcb35e53860083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010d47757cecd6b4fc6ef7f12206d1118ea08f104607954e8f0089085028672c5c0f50083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010e98923edb26908ea72859d8ef8894b7408f120e429f81109645ad65b77fa2df601b408e097fa6a4a4f668431dd23da0b189e5208f104607954e9f008689b099f717513f10083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010fa58b5c61423f4cdacaafb917c817d7708f1046079470ef008915a75964c3c57340083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

**Comment by ajtowns** [comment]
Post-merge ACK cbd64c3a28a7466f421477daadc6e6e6b69b898a -- checked fuzzer compiles at each step, visually compared new patches with originals, checked original patches are in master

**Pull request merged**